### ISSUE TITLE
libpam 1.4.0 makes pam_usb unusable

### ISSUE THREAD

**Issue by KeyofBlueS** [mcdope/pam_usb#36]
After today's upgrade (debian sid) of libpam-modules libpam-modules-bin libpam-runtime libpam0g from 1.3.1-5 to 1.4.0-1, pam_usb stopped working, as if it wasn't installed at all. Reverting them back to 1.3.1-5 solves the issue.

**Comment by mcdope** [comment]
It *shouldn't* affect it based on the package changelog. Will setup a sid VM to test.

Could you please check if you are maybe just affected by the `securetty` / `nullok_secure` removal? See https://bugs.launchpad.net/ubuntu/+source/pam/+bug/1860826/comments/15 for details.

**Comment by mcdope** [comment]
I've installed a sid minimal VM, only additional stuff selected was "Standard Systemwerkzeuge" (Default system utilities or similar on english) and it worked fine.

Install and check for user (from root context)
![image](https://user-images.githubusercontent.com/3312979/103297957-a773ca80-49f9-11eb-8f6b-4d9e545affd9.png)

check for user (from user context)
![image](https://user-images.githubusercontent.com/3312979/103298070-e275fe00-49f9-11eb-9b6d-28cc65d132bb.png)

**Comment by KeyofBlueS** [comment]
I've changed `nullok_secure` with `nullok` but this didn't solve the issue.

Have you upgraded your system after installation? What libpam version do you have?

**Comment by mcdope** [comment]
Yes and no, I've installed using the ISO from 2020-12-28 and installed all available upgrades. But no, I haven't upgraded it really in the sense of earlier package oder release versions.

Just tested if I can login in that VM using USB, and yes - I can. It's not only check that's working here.

![Peek 2020-12-29 18-08](https://user-images.githubusercontent.com/3312979/103301258-4b14a900-4a01-11eb-957a-f8c09d0057cb.gif)


Would be nice if you could enable debug globally, in `/etc/security/pam_usb.conf` so we get the actual debug output of the pam usage. But if check works, login should work too.

**Comment by KeyofBlueS** [comment]
Enabling debug globally gives no additional info

**Comment by mcdope** [comment]
So even with `<debug>true</debug>` you get no output at all? Oo

Sounds like the module isn't invoked at all, that's really weird. Will try to reproduce that further tomorrow. Maybe downgrading package and then upgrading again will reproduce it.

**Comment by KeyofBlueS** [comment]
No, nothing at all.
I'll keep trying and searching...

**Comment by KeyofBlueS** [comment]
`pamusb-agent` seems ok (even tried to stop and start it):
```
user@debian:~$ sudo systemctl status pamusb-agent
● pamusb-agent.service - USB PAM authentication agent
     Loaded: loaded (/lib/systemd/system/pamusb-agent.service; enabled; vendor preset: enabled)
     Active: active (running) since Tue 2020-12-29 23:08:27 CET; 1min 4s ago
   Main PID: 4218 (python3)
      Tasks: 5 (limit: 9496)
     Memory: 9.6M
        CPU: 131ms
     CGroup: /system.slice/pamusb-agent.service
             └─4218 python3 /usr/bin/pamusb-agent

dic 29 23:08:27 debian systemd[1]: Started USB PAM authentication agent.
dic 29 23:08:28 debian pamusb-agent[4218]: pamusb-agent[4218]: pamusb-agent up and running.
dic 29 23:08:28 debian pamusb-agent[4218]: pamusb-agent up and running.
dic 29 23:08:28 debian pamusb-agent[4218]: Watching device "test" for user "root"
dic 29 23:08:28 debian pamusb-agent[4218]: pamusb-agent[4218]: Watching device "test" for user "root"
dic 29 23:08:28 debian pamusb-agent[4218]: pamusb-agent[4218]: Watching device "test" for user "user"
dic 29 23:08:28 debian pamusb-agent[4218]: Watching device "test" for user "user"
```
 I've checked everything I can think of, but I don't know what else to do, just revert to 1.3.1-5 works.

**Comment by mcdope** [comment]
Created a restore point for my buster VM and upgrading it to unstable right now. Hopefully the bug will occur.

**Comment by mcdope** [comment]
Managed to reproduce it. But most likely I will have no further time to investigate it this year. Will have to wait till next year. 

But will prepare 0.7.1 with the other fixes tomorrow morning.

**Comment by mcdope** [comment]
What da... Found this in `/var/log/auth.log`:

![image](https://user-images.githubusercontent.com/3312979/103366341-ac508100-4ac2-11eb-9ab5-6a9445f41b57.png)

Either one of the pam packages removes stuff it isn't supposed to remove (unlikely) or we install into some wrong location. Seems our `.so` file gets deleted on upgrade oO. This definitely needs some deeper investigation.

**Comment by KeyofBlueS** [comment]
No! The rest of the 2020 without a fix? XD

I tried to reinstall libpam-usb AFTER the upgrade, should this reinstall pam_usb.so if it was removed?

**Comment by KeyofBlueS** [comment]
sudo ln -s /lib/security/pam_usb.so /lib/x86_64-linux-gnu/security/pam_usb.so

This solves the issue

**Comment by mcdope** [comment]
Then we are installing into the wrong directory for x64 it seems. 

I've seen stuff related to this in other forks, think I can fix that without much effort. But still, 2021 it is ^^

**Comment by miwie** [comment]
Can I expect any improvement on this issue before I give pam_usb a try on my Debian testing system?

**Comment by mcdope** [comment]
You can expect a fix, yes. But I can't say if this will be before you try it :D 

I will take a look at some point in this week, but having a day-time job plus my own business so there are some time issues involved ;-)

**Comment by miwie** [comment]
FYI: 
- changed `PAM_USB_DEST` to `$(DESTDIR)/$(LIBDIR)/x86_64-linux-gnu/security` in `Makefile`
- `make deb`
- verified the new location of `pam_usb.so` in the deb file: `/lib/x86_64-linux-gnu/security/pam_usb.so`
- installed/configured _pam_usb_ on a Debian _Testing/Bullseye_ system
- works as expected!

I'm not sure whether this (or a similar) change in `Makefile` is sufficient. There might be additional changes necessary depending on the architecture?

**Comment by mcdope** [comment]
Anyone want to test the fix?

https://www.dropbox.com/s/i9wv9wzg6m1lsn4/libpam-usb_0.7.1_amd64.deb?dl=0

**Comment by miwie** [comment]
Tested and looks like it works w/o problems.

**Comment by mcdope** [comment]
Thanks, saved me some VM action :)

Will merge later this day then.

**Comment by mcdope** [comment]
Fix released as 0.7.2, repo is updated.

https://apt.mcdope.org/libpam-usb_0.7.2_amd64.deb
